### PR TITLE
[Merged by Bors] - feat: #leansearch command

### PR DESCRIPTION
**Pull request by siddhartha-gadgil**
We add a dependenct to `LeanSearchClient`, which provides syntax for search using the [leansearch API](https://leansearch.net/) from within Lean. This allows search for Lean tactics and theorems using natural language.

LeanSearchClient provide syntax to make a query and generate `TryThis` options to click or use a code action to use the results. The queries are of three forms:

* `Command` syntax: `#leansearch "search query"` as a command.
* `Term` syntax: `#leansearch "search query"` as a term.
* `Tactic` syntax: `#leansearch "search query"` as a tactic.

In all cases results are displayed in the Lean Infoview and clicking these replaces the query text. In the cases of a query for tact
ics only valid tactics are displayed.

## Examples

The following are examples of using the leansearch API. The search is triggered when the sentence ends with a full stop (period) or 
a question mark.

### Query Command

```lean
#leansearch "If a natural number n is less than m, then the successor of n is less than the successor of m"
```

### Query Term

```lean
example := #leansearch "If a natural number n is less than m, then the successor of n is less than the successor of m"
```

### Query Tactic

Note that only valid tactics are displayed.

```lean
example : 3 ≤ 5 := by
  #leansearch "If a natural number n is less than m, then the successor of n is less than the successor of m"
```

### CODE REVIEW

File: Mathlib/Util/LeanSearchExamples.lean [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/-
+Copyright (c) 2024 Siddhartha Gadgil. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddhartha Gadgil
+-/
```
Comment (semorrison) [inline Mathlib/Util/LeanSearchExamples.lean:5]: Please move this file to `tests`.

File: Mathlib/Util/LeanSearchExamples.lean [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import Lean
+import LeanSearchClient.Syntax
```
Comment (semorrison) [inline Mathlib/Util/LeanSearchExamples.lean:7]:
```suggestion
import LeanSearchClient
```

File: Mathlib/Util/LeanSearchExamples.lean [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/-
+Copyright (c) 2024 Siddhartha Gadgil. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddhartha Gadgil
+-/
+import Lean
```
Comment (semorrison) [inline Mathlib/Util/LeanSearchExamples.lean:6]: Is this needed?

File: Mathlib/Util/LeanSearchExamples.lean (hunk 1 above)
Comment (siddhartha-gadgil) [reply Mathlib/Util/LeanSearchExamples.lean]: Done. Moved to `test/LeanSearchExamples.lean`

File: Mathlib/Util/LeanSearchExamples.lean (hunk 2 above)
Comment (siddhartha-gadgil) [reply Mathlib/Util/LeanSearchExamples.lean]: Made the change (after moving).

File: Mathlib/Util/LeanSearchExamples.lean (hunk 3 above)
Comment (siddhartha-gadgil) [reply Mathlib/Util/LeanSearchExamples.lean]: It is not needed. I removed it.

File: lakefile.lean [hunk 4]
```diff
@@ -12,6 +12,8 @@ require "leanprover-community" / "Qq" @ git "master"
 require "leanprover-community" / "aesop" @ git "master"
 require "leanprover-community" / "proofwidgets" @ git "v0.0.42"
 require "leanprover-community" / "importGraph" @ git "main"
+require "LeanSearchClient" from git
+  "https://github.com/siddhartha-gadgil/LeanSearchClient.git" @ "main"
```
Comment (semorrison) [inline lakefile.lean:16]: One line, please

Comment (semorrison) [inline lakefile.lean:16]:
```suggestion
require "LeanSearchClient" from git "https://github.com/siddhartha-gadgil/LeanSearchClient.git" @ "main"
```

File: lakefile.lean [hunk 5]
```diff
@@ -12,6 +12,7 @@ require "leanprover-community" / "Qq" @ git "master"
 require "leanprover-community" / "aesop" @ git "master"
 require "leanprover-community" / "proofwidgets" @ git "v0.0.42"
 require "leanprover-community" / "importGraph" @ git "main"
+require "LeanSearchClient" from git "https://github.com/siddhartha-gadgil/LeanSearchClient.git" @ "main"
```
Comment (semorrison) [inline lakefile.lean:15]:
```suggestion
require "siddhartha-gadgil" / "LeanSearchClient" @ git "main"
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**